staging/src/k8s.io/apiserver/pkg/storage/cacher/watch_cache.go

```diff
@@ -545,15 +554,45 @@ func (w *watchCache) waitUntilFreshAndListItems(ctx context.Context, resourceVer
 		// TODO: if multiple indexes match, return the one with the fewest items, so as to do as much filtering as possible.
 		for _, matchValue := range matchValues {
 			if result, err := w.store.ByIndex(matchValue.IndexName, matchValue.Value); err == nil {
+				result, err = filterPrefix(key, result)
+				if err != nil {
+					return nil, 0, "", err
+				}
+				sort.Sort(sortableStoreElements(result))
```

Comment:
Is there a reason the store implementations are not responsible for this?

Reply:
No, other than we are not the owners of storeIndexer, as it lies in client library. That would mean we would need to copy out the storage to watch cacher package.

Reply:
Also having filtering before sorting is a memory/allocation optimization, as this is a critical code path I think we would want to leave it as it is, or we would need to make store aware of filtering.